securesystemslib/signer/_signer.py

```diff
@@ -0,0 +1,264 @@
+
+    @abc.abstractmethod
+    def sign(self, payload: bytes) -> Signature:
+        """Signs a given payload by the key assigned to the Signer instance.
+
+        Arguments:
+            payload: The bytes to be signed.
+
+        Returns:
+            Returns a "Signature" class instance.
+        """
+        raise NotImplementedError  # pragma: no cover
+
+    @classmethod
+    @abc.abstractmethod
+    def from_priv_key_uri(
+        cls,
+        priv_key_uri: str,
+        public_key: Key,
+        secrets_handler: Optional[SecretsHandler] = None,
+    ) -> "Signer":
+        """Factory constructor for a given private key URI
+
+        Returns a specific Signer instance based on the private key URI and the
+        supported uri schemes listed in SIGNER_FOR_URI_SCHEME.
+
+        Args:
+            priv_key_uri: URI that identifies the private key
+            public_key: Key that is the public portion of this private key
+            secrets_handler: Optional function that may be called if the
+                signer needs additional secrets (like a PIN or passphrase).
+                secrets_handler should return the requested secret string.
+
+        Raises:
+            ValueError: Incorrect arguments
+            Other Signer-specific errors: These could include OSErrors for
+                reading files or network errors for connecting to a KMS.
+        """
+
+        scheme, _, _ = priv_key_uri.partition(":")
+        if scheme not in SIGNER_FOR_URI_SCHEME:
+            raise ValueError(f"Unsupported private key scheme {scheme}")
+
+        signer = SIGNER_FOR_URI_SCHEME[scheme]
+        return signer.from_priv_key_uri(
+            priv_key_uri, public_key, secrets_handler
+        )
+
+
+class SSlibSigner(Signer):
+    """A securesystemslib signer implementation.
+
+    Provides a sign method to generate a cryptographic signature with a
+    securesystemslib-style rsa, ed25519 or ecdsa key. See keys module
+    for the supported types, schemes and hash algorithms.
+
+    SSlibSigners should be instantiated with Signer.from_priv_key_uri().
+    These private key URI schemes are supported:
+    * "envvar:<VAR>":
+        VAR is an environment variable with unencrypted private key content.
+           envvar:MYPRIVKEY
+    * "file:<PATH>?encrypted=[true|false]":
+        PATH is a file path to a file with unencrypted private key content.
+           file:path/to/file?encrypted=true
+           file:/abs/path/to/file?encrypted=false
+
+    Attributes:
+        key_dict:
+            A securesystemslib-style key dictionary. This is an implementation
+            detail, not part of public API
+    """
+
+    ENVVAR_URI_SCHEME = "envvar"
+    FILE_URI_SCHEME = "file"
+
+    def __init__(self, key_dict: Dict):
+        self.key_dict = key_dict
+
+    @classmethod
+    def from_priv_key_uri(
+        cls,
+        priv_key_uri: str,
+        public_key: Key,
+        secrets_handler: Optional[SecretsHandler] = None,
```

Comment:
There is no documentation for the `secrets_handler` argument.

Reply:
Good catch. I'm torn about documenting these methods at all. The abstractmethod is already documented in the base Signer: copy-pasting to SSlibSigner seems silly... but on the other hand there would be the additional detail like specific errors raised by sslibsigner.... and on the third hand no-one is going to read these specific Signer implementations docs since their code will be using the Signer class -- they won't even know which implementation will be dispatched.

The underlying issue is that because  SSlibSigner is just an implementation of Signer _it should not need any new method docs_. But for that to be 100% true, I would have to add a "SignerError" that catches all errors from the dispatched implementation (otherwise OSErrors or google.cloud errors are bound to happen)... and that just seems wrong.

I think this might need some more thought. A good example situation is a GCPSigner failing because there is no network connection... This could be a case that a repository application wants to handle. But I also think we could accept the situation for now and try to come up with a solution before this is released

Reply:
Yes, there's an unresolvable conflict: 
- users should only know the base class
- users should know traits that are specific to subclasses (exceptions, name of a secret)

I think there are only two ways to overcome his:
- discard one of the assumptions (probably the former)
- make all subclass traits available via the base class (that does not seem desirable or feasible)

Reply:
Ok, I've mostly removed the docs for SSlibSigner: I think copy-pasting is not useful considering the users will be reading Signer docs anyway...

But I have added a TODO item and a new comment about the exception issue that this revealed.

resolving this conversation